src/nvim/event/stream.c

```diff
@@ -100,6 +102,8 @@ void stream_may_close(Stream *stream)
 void stream_close_handle(Stream *stream)
   FUNC_ATTR_NONNULL_ALL
 {
+  assert(!stream->uv_closed);
+  stream->uv_closed = true;
```

Comment:
What's going on here? We are tracking two different closed states?

Reply:
Deadlock on exiting is what is going on. `closed` appearently doesn't mean "closed" always, only "intent on close" as soon as the internal refcount goes to zero. But in mch_exit we cannot wait, we must close everything _now_ (and we won't return to the event handler holding the refcount anyway).

Alternatively we could refactor exiting to not invoke arbitrary recursive event handling (a more "do what's really necessary and then exit(3)" style of exiting) but that's probably another PR...

Reply:
The situation occurs because "press RETURN". We "formally" forbid event processing at press-RETURN, so it "safely" can happen even in any event handler, but that's very much a false promise, as nvim could always receive an exiting signal at press-RETURN, which as mentioned, invokes arbitrary event processing...